kernel/src/table.rs

```diff
@@ -46,6 +83,46 @@ impl Table {
     }
 }
 
+#[derive(Debug)]
+enum UriType {
+    LocalPath(PathBuf),
+    Url(Url),
+}
+
+/// Utility function to figure out whether string representation of the path is either local path or
+/// some kind or URL.
+///
+/// Will return an error if the path is not valid.
+fn resolve_uri_type(table_uri: impl AsRef<str>) -> DeltaResult<UriType> {
+    let table_uri = table_uri.as_ref();
+    let known_schemes = [
+        "file", "memory", "s3", "s3a", "az", "adl", "azure", "abfs", "abfss", "gs", "http", "https",
+    ]; // hard-code for now, TODO: pull these out of object_store
+
+    if let Ok(url) = Url::parse(table_uri) {
+        let scheme = url.scheme().to_string();
+        if url.scheme() == "file" {
+            Ok(UriType::LocalPath(
+                url.to_file_path()
+                    .map_err(|_| Error::invalid_table_location(table_uri))?,
+            ))
+        } else if known_schemes.contains(&scheme.as_str()) {
+            Ok(UriType::Url(url))
+        // NOTE this check is required to support absolute windows paths which may properly parse as url
+        // we assume here that a single character scheme is a windows drive letter
+        } else if scheme.len() == 1 {
+            Ok(UriType::LocalPath(PathBuf::from(table_uri)))
+        } else {
+            Err(Error::InvalidTableLocation(format!(
+                "Unknown scheme: {}",
+                scheme
+            )))
+        }
```

Comment:
in delta-rs we specified these known schemes as they are the ones our storage backends can deal with - ultimately those we support in upstream `object_store`.

Since we technically have no opinion on what a valid scheme is - some engines might even accept `dbfs` :) - do we want to fail on unknown schemes?